**Тема “Обучение с учителем”**

Задание 1

Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.


In [2]:
import numpy as np
import pandas as pd

In [3]:
from sklearn.datasets import load_boston

In [ ]:
boston = load_boston()
data = boston.data
target = boston.target
feature_names = boston.feature_names

In [5]:
x = pd.DataFrame(data, columns = feature_names)
x.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [6]:
x.memory_usage()

Index       128
CRIM       4048
ZN         4048
INDUS      4048
CHAS       4048
NOX        4048
RM         4048
AGE        4048
DIS        4048
RAD        4048
TAX        4048
PTRATIO    4048
B          4048
LSTAT      4048
dtype: int64

In [7]:
#Потренировалась с сокращением памяти:)

def reduce_mem_usage(x):
  start_mem = x.memory_usage().sum() / 1024**2
  print('Memory usage of dataframe is {:.2f} MB'.format(start_mem))
  
  for col in x.columns:
    col_type = x[col].dtype
    
    if col_type != object:
      c_min = x[col].min()
      c_max = x[col].max()
      if str(col_type)[:3] == 'int':
        if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
          x[col] = x[col].astype(np.int8)
        elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
          x[col] = x[col].astype(np.int16)
        elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
          x[col] = x[col].astype(np.int32)
        elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
          x[col] = x[col].astype(np.int64)
      else:
        if c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
          x[col] = x[col].astype(np.float32)     
        else:
          x[col] = x[col].astype(np.float64)
    else:
      x[col] = x[col].astype('category')
  
  end_mem = x.memory_usage().sum() / 1024**2
  print('Memory usage after optimization is: {:.2f} MB'.format(end_mem))
  print('Decreased by {:.1f}%'.format(100 * (start_mem - end_mem) / start_mem))
  
  return x

In [8]:
x = reduce_mem_usage(x)

Memory usage of dataframe is 0.05 MB
Memory usage after optimization is: 0.03 MB
Decreased by 49.9%


In [9]:
y = pd.DataFrame(target, columns = ['price'])
y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

In [10]:
from sklearn.model_selection import train_test_split

In [11]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 42)

In [12]:
from sklearn.linear_model import LinearRegression

In [13]:
lr = LinearRegression()
lr.fit(x_train, y_train)

LinearRegression()

In [14]:
y_pred = lr.predict(x_test)

In [15]:
check_test = pd.DataFrame({'y_test': y_test['price'],
                           'y_pred': y_pred.flatten()},
                          columns = ['y_test', 'y_pred'])
check_test.head(10)

y_test     y_pred
173    23.6  28.648968
274    32.4  36.495007
491    13.6  15.411242
72     22.8  25.403122
452    16.1  18.855324
76     20.0  23.146690
316    17.8  17.392132
140    14.0  14.078630
471    19.6  23.036983
500    16.8  20.599442

In [16]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
mean_squared_error(y_test, y_pred)

21.517450546419962

In [17]:
mean_absolute_error(y_test, y_pred)

3.1627087731110417

In [18]:
r2_score(y_test, y_pred)

0.711225920995032

Задание 2

Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble. Сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42. Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy, так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма. Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания. Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [19]:
from sklearn.ensemble import RandomForestRegressor

In [31]:
model = RandomForestRegressor(n_estimators = 1000,
                              max_depth = 12,
                              random_state=42)

In [32]:
model.fit(x_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [34]:
y_pred = model.predict(x_test)

In [35]:
check_test = pd.DataFrame({'y_test': y_test['price'],
                           'y_pred': y_pred.flatten()},
                          columns = ['y_test', 'y_pred'])
check_test.head(10)

y_test     y_pred
173    23.6  22.806412
274    32.4  31.131464
491    13.6  16.339125
72     22.8  23.810726
452    16.1  17.139521
76     20.0  21.832284
316    17.8  19.895747
140    14.0  14.754118
471    19.6  21.240835
500    16.8  20.898658

In [36]:
r2_score(y_test, y_pred)

0.87472606157312

R2 из предыдущего задания 0.711225920995032, т.к. R2 из второго задания ближе к единице, модель случайного леса работает в данном случае лучше

Задание 3

Вызовите документацию для класса RandomForestRegressor, найдите информацию об атрибуте feature_importances_. С помощью этого атрибута найдите сумму всех показателей важности, установите, какие два признака показывают наибольшую важность.

In [37]:
model.feature_importances_

array([0.03167574, 0.00154252, 0.00713813, 0.00123624, 0.01426897,
       0.40268179, 0.01429864, 0.06397257, 0.00528122, 0.01152493,
       0.01808108, 0.01245085, 0.41584732])

In [40]:
np.sum(model.feature_importances_)

1.0

In [45]:
imp = pd.DataFrame({'name': x.columns,
                    'feature_importance': model.feature_importances_},
                   columns = ['feature_importance', 'name'])
imp.nlargest(2, 'feature_importance')

feature_importance   name
12            0.415847  LSTAT
5             0.402682     RM